stdlib/strtod.c

```diff
@@ -15,184 +15,539 @@
+	}
+	else {
+		str = scan_number_long_double(str, hex, LDBL_MANT_10_DIG, &frac_position, &overflow_int, &result);
+	}
+
+	if (frac_position > (INT32_MAX / 4) || overflow_int > (INT32_MAX / 4)) {
+		/* This is an absurd situation, we read in 2^29 digits from input, so we don't care much about the result */
+		total_exp = (frac_position != 0) ? INT32_MIN : INT32_MAX;
+	}
+	else {
+		/* If there was a decimal point, we don't need to care about the overflow */
+		total_exp = (frac_position != 0) ? -(int64_t)frac_position : overflow_int;
+		if (hex) {
+			total_exp *= 4;
+		}
 	}
 
-	long double result = whole_number;
-	result += fraction * pow(10, -dot_position);
 
-	tmp = base;
-	for (i = 0; i < powers; i++) {
-		if ((exp >> i) & 1) {
-			if (exp_negative != 0) {
-				result /= tmp;
+	bool has_exponent = hex ? (*str == 'P' || *str == 'p') : (*str == 'E' || *str == 'e');
+	if (has_exponent) {
+		const char *str_exp = str + 1;
+		bool exp_negative = false;
+		if ((*str_exp == '-') || (*str_exp == '+')) {
+			exp_negative = (*str_exp == '-');
+			str_exp++;
+		}
+
+		if (isdigit(*str_exp)) {
+			size_t exp_overflowed = 0, dummy;
+			uint64_t exp_scan;
+			str = scan_number(str_exp, 0, false, ((INT32_MAX - 9) / 10), &dummy, &exp_overflowed, &exp_scan);
+			if (exp_overflowed != 0) {
+				total_exp = exp_negative ? INT32_MIN : INT32_MAX;
 			}
 			else {
-				result *= tmp;
+				/* At this point exp_scan is guaranteed to be <= INT32_MAX */
+				total_exp += exp_negative ? -(int32_t)exp_scan : exp_scan;
 			}
+		}
+	}
 
-			is_number = 1;
+	if (result != 0) {
+		if (total_exp < exp_min) {
+			result = 0;
+		}
+		else if (total_exp > exp_max) {
+			result = INFINITY;
+		}
+		else if (hex) {
+			result = mul_exp2(result, total_exp);
+		}
+		else {
+			result = mul_exp10(result, total_exp);
 		}
 
-		tmp *= tmp;
+		if (result == 0) {
+			/* We entered with non-zero result, but now it is zero - underflow must have occurred */
+			/* TODO: detect denormalization */
+			errno = ERANGE;
+		}
 	}
 
-	if (negative != 0) {
-		result = -result;
+	*out = result;
+	return str;
+}
+
+
+static long double strto_any(const char *__restrict str, char **__restrict endptr, uint8_t type)
+{
+	int sign;
+	long double result = 0.0l;
+
+	if (str == NULL) {
+		return 0.0l;
 	}
 
-	if (endptr != NULL) {
-		*endptr = (char *)str;
+	const char *str_wsin = floatparse_wsin(str, &sign, &result);
+	if (sign == 0) {
+		if (endptr != NULL) {
+			*endptr = (char *)str_wsin;
+		}
+
+		return result;
 	}
 
-	if (!is_number) {
+	bool is_hex;
+	const char *str_zerox = floatparse_zerox(str_wsin, &is_hex);
+	if (str_zerox == NULL) {
 		errno = EINVAL;
+		if (endptr != NULL) {
+			*endptr = (char *)str;
+		}
+
+		return result;
 	}
-	else if (isinf(result)) {
+
+	str = floatparse_number(str_zerox, is_hex, type, &result);
+
+	bool range_error;
+	switch (type) {
+		case TYPE_ID_FLT: {
+			float cast = (float)result;
+			range_error = isinf(cast) || ((result != 0) && (cast == 0));
```

Comment:
nitpick:
```suggestion
			range_error = isinf(cast) || ((result != 0.0l) && (cast == 0.0f));
```
Applies to everywhere. Makes variables more "self-documenting".

Reply:
Done